Add integration tests for resource types with parameters

jghiloni opened the pull request:
This adds integration tests for https://github.com/concourse/atc/pull/220, which was in response to https://github.com/concourse/concourse/issues/1690

vito commented on pipelines/fixtures/resource-types-with-params.yml:
```diff
@@ -0,0 +1,93 @@
+    uri: {{origin-git-server}}
+    branch: master
+  params:
+    foo: bar
```
Is this actually observed/verified in the tests somehow? In other words, would these tests fail without the implementation, or would this just be silently ignored?

This may be difficult to verify, to be honest; the [git resource params](https://github.com/concourse/git-resource#parameters) only has a few params that we could really detect being used, and even then, this only applies to the resource type fetching...

Maybe we could do something wacky like point it to a repo containing some of the resource type's code as a submodule? But it even defaults to fetching them by default, so we could only really turn it off and see that it fails when the params are respected. Weird stuff.

jghiloni commented:
The easiest way to verify this would be to have a resource type with a
rootfs stored in s3. Setting
params: {unpack: true} would allow it to work yes?

On Mon, Nov 13, 2017 at 16:24 Alex Suraci <notifications@github.com> wrote:

> *@vito* requested changes on this pull request.
> ------------------------------
>
> In pipelines/fixtures/resource-types-with-params.yml
> <https://github.com/concourse/testflight/pull/3#discussion_r150669270>:
>
> > @@ -0,0 +1,93 @@
> +---
> +resource_types:
> +- name: custom-resource-type
> +  type: git
> +  privileged: ((privileged))
> +  source:
> +    uri: {{origin-git-server}}
> +    branch: master
> +  params:
> +    foo: bar
>
> Is this actually observed/verified in the tests somehow? In other words,
> would these tests fail without the implementation, or would this just be
> silently ignored?
>
> This may be difficult to verify, to be honest; the git resource params
> <https://github.com/concourse/git-resource#parameters> only has a few
> params that we could really detect being used, and even then, this only
> applies to the resource type fetching...
>
> Maybe we could do something wacky like point it to a repo containing some
> of the resource type's code as a submodule? But it even defaults to
> fetching them by default, so we could only really turn it off and see that
> it fails when the params are respected. Weird stuff.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/concourse/testflight/pull/3#pullrequestreview-76254344>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABv9GCzNeS_KvNrhNsMvBXhBvqK5Jg43ks5s2LOggaJpZM4QaIzn>
> .
>
-- 

Josh Ghiloni
jghiloni@pivotal.io

jghiloni commented:
I know we talked about using the `time-resource` but I couldn't get it to effectively run in test-flight as it wouldn't recognize new versions. I switched to `semver-resource` so that it would work.

vito commented:
@jghiloni That's weird. If configured with `interval` it should return a version right off the bat. What happened?

jghiloni commented:
IIRC, it was returning that there were versions available, but the job wasn't recognizing them. The semver works nicely though, and isn't TOO much larger of an image (50MB vs 20MB)

vito commented:
Not sure what you mean by not recognizing them. Did the `TriggerJob` just sit there and cause the test to hang?

I just want to make sure to understand this since to me using semver for this seems a bit overkill. Not really worried about the size, just the semantics are a bit confusing for the test.

Could you also include scripts that set up the bucket that it references? Then maybe I can try it out myself. I want to block the PR on that anyway just to make sure everything is reproducible.

Thanks!

jghiloni commented:
I switched back to time after resolving the issues that were there (combination of PEBKAC and existing issues around radar scanning of resource types). As for the script that created the bucket, I don't have one but I can put one together real quick.

vito commented on pipelines/fixtures/resource-types-with-params.yml:
```diff
@@ -0,0 +1,31 @@
+- name: time-from-s3
+  type: s3
+  source:
+    bucket: {{s3_bucket}}
```
let's use the newer `((params))` syntax instead

vito commented on pipelines/resource_type_test.go:
```diff
@@ -131,6 +131,26 @@ var _ = Describe("Configuring a resource type in a pipeline config", func() {
+				pipelineName,
+				"-c", "fixtures/resource-types-with-params.yml",
+				"-v", "origin-git-server="+originGitServer.URI(),
+				"-v", "s3_bucket=jghiloni-psrtest",
```
this is fine for now; once i re-create the bucket with the script you provide, i'll update it during the merge

jghiloni commented:
Ok @vito this should take care of everything you've asked for. I switched to using `((foo))` instead of `{{foo}}`, and added a script to upload a rootfs to a new s3 bucket (with a concourse task to generate the rootfs from a docker image, if necessary)

jghiloni commented:
Hi @vito et al, is there a chance that this and https://github.com/concourse/atc/pull/220 could get an up or down vote today? It'll make our client's life a lot easier if we have it.

vito commented on pipelines/resource_type_test.go:
```diff
@@ -131,6 +132,26 @@ var _ = Describe("Configuring a resource type in a pipeline config", func() {
+		})
+
+		It("can use a custom resource with parameters", func() {
+			time.Sleep(30 * time.Second)
```
Is this sleep still necessary now that the task uses a `docker-image` `image_resource`?

jghiloni replied:
Yeah, it is for some reason. The resource goes orange until some time has passed, and adding the sleep back in makes the tests pass.

vito commented:
Takin' a look at this now. We had some work in flight in the radar that might remove the need for the sleep. I'll remove it and merge if so.

vito commented:
merged manually (and applied the sleep fix myself)!
